libbeat/processors/xmldecode/xmldecode.go

```diff
@@ -0,0 +1,162 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package xmldecode
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/beat/events"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/jsontransform"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	"github.com/elastic/beats/v7/libbeat/processors/checks"
+	jsprocessor "github.com/elastic/beats/v7/libbeat/processors/script/javascript/module/processor"
+)
+
+type xmlDecodeConfig struct {
+	Fields        []string `config:"fields"`
+	ExpandKeys    bool     `config:"expand_keys"`
+	OverwriteKeys bool     `config:"overwrite_keys"`
+	AddErrorKey   bool     `config:"add_error_key"`
+	Target        *string  `config:"target"`
+	DocumentID    string   `config:"document_id"`
+	ToLower       bool     `config:"to_lower"`
+}
+
+type xmlDecode struct {
+	config xmlDecodeConfig
+	logger *logp.Logger
+}
+
+var (
+	defaultConfig = xmlDecodeConfig{
+		Fields:        []string{"message"},
+		OverwriteKeys: false,
+		ExpandKeys:    false,
+		AddErrorKey:   false,
+		ToLower:       true,
+	}
+)
+
+func init() {
+	processors.RegisterPlugin("xmldecode",
+		checks.ConfigChecked(NewDecodeXMLFields,
+			checks.RequireFields("fields"),
+			checks.AllowedFields("fields", "overwrite_keys", "add_error_key", "expand_keys", "target", "document_id")))
+	jsprocessor.RegisterPlugin("XMLDecode", NewDecodeXMLFields)
+}
+
+// NewDecodeXMLFields construct a new decode_xml_fields processor.
+func NewDecodeXMLFields(c *common.Config) (processors.Processor, error) {
+	config := defaultConfig
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("fail to unpack the xmldecode configuration: %s", err)
+	}
+
+	return &xmlDecode{
+		config: config,
+		logger: logp.NewLogger("xmldecode"),
+	}, nil
+
+}
+
+func (x *xmlDecode) Run(event *beat.Event) (*beat.Event, error) {
+	var errs []string
+
+	for _, field := range x.config.Fields {
+		data, err := event.GetValue(field)
+		if err != nil && errors.Cause(err) != common.ErrKeyNotFound {
+			x.logger.Debugf("Error trying to GetValue for field : %s in event : %v", field, event)
+			errs = append(errs, err.Error())
+			continue
+		}
+
+		xmloutput, err := x.decodeField(field, data)
```

Comment:
This part of the code is exactly how` decode_json_fields` performs the action though @jsoriano ? Is it not so that it does not completely fail if one of multiple fields does not exist?

Reply:
Before unmarshalling, `decode_json_fields` checks if the data is actually a string: https://github.com/elastic/beats/blob/3e47a781bb0e11a04caed9ab87cad5ad6ea80350/libbeat/processors/actions/decode_json_fields.go#L107-L121

This effectively prevents trying to decode `null` if `err == common.ErrKeyNotFound`, and is in general safer in case `data` is of any other non-string type.

With this you would also have in `text` a string, so you wouldn't need the `Sprintf` in https://github.com/elastic/beats/pull/23678/files#diff-90271196007e26ceaf04e257f546933b4a3c553145831bfd695559bd4bbc1ef5R145.